dm/pkg/binlog/pos_finder.go

```diff
@@ -0,0 +1,370 @@
+
+		r.tctx.L().Info("min timestamp in binlog file", zap.Reflect("file", currBinlog), zap.Uint32("ts", minTs))
+
+		lastTs = minTs
+		lastMid = mid
+
+		if minTs >= targetTs {
+			end = mid - 1
+		} else {
+			// current binlog maybe the target binlog file, we'll backtrace to it later.
+			begin = mid + 1
+		}
+	}
+	if lastTs >= targetTs {
+		if lastMid == 0 {
+			// timestamp of first binlog event in first binlog file >= targetTs
+			r.targetBinlog = binaryLogs[lastMid]
+			r.tsBeforeFirstBinlog = true
+		} else {
+			// timestamp of first event in lastMid >= targetTs, need to search from previous binlog file
+			r.targetBinlog = binaryLogs[lastMid-1]
+		}
+	} else {
+		r.targetBinlog = binaryLogs[lastMid]
+	}
+	r.lastBinlogFile = r.targetBinlog.name == binaryLogs[len(binaryLogs)-1].name
+	return nil
+}
+
+func (r *binlogPosFinder) processGTIDRelatedEvent(ev *replication.BinlogEvent, prevSet gtid.Set) (gtid.Set, error) {
+	ev, err := r.parser.Parse(ev.RawData)
+	if err != nil {
+		return nil, err
+	}
+	switch ev.Header.EventType {
+	case replication.PREVIOUS_GTIDS_EVENT:
+		newSet, err := event.GTIDsFromPreviousGTIDsEvent(ev)
+		if err != nil {
+			return nil, err
+		}
+		return newSet, nil
+	case replication.MARIADB_GTID_EVENT:
+		fallthrough
+	case replication.GTID_EVENT:
+		gtidStr, _ := event.GetGTIDStr(ev)
+		if err := prevSet.Update(gtidStr); err != nil {
+			return nil, err
+		}
+	case replication.MARIADB_GTID_LIST_EVENT:
+		newSet, err := event.GTIDsFromMariaDBGTIDListEvent(ev)
+		if err != nil {
+			return nil, err
+		}
+		return newSet, nil
+	}
+	return prevSet, nil
+}
+
+func (r *binlogPosFinder) checkTransactionBoundaryEvent(ev *replication.BinlogEvent) (bool, error) {
+	// we find the timestamp at transaction boundary
+	// When there are GTID events in this binlog file, we use GTID event as the start event, else:
+	// for DML
+	// 		take a 'BEGIN' query event as the start event
+	//		XID event or a 'COMMIT' query event as the end event
+	// for DDL
+	// 		one single query event acts as both the start and end event
+	var transactionBeginEvent bool
+	switch ev.Header.EventType {
+	case replication.FORMAT_DESCRIPTION_EVENT:
+		_, err := r.parser.Parse(ev.RawData)
+		if err != nil {
+			return false, err
+		}
+	case replication.GTID_EVENT:
+		fallthrough
+	case replication.ANONYMOUS_GTID_EVENT: // since 5.7, when GTID not enabled. we use this to avoid parsing query event
+		fallthrough
+	case replication.MARIADB_GTID_EVENT:
+		r.everMetGTIDEvent = true
+		transactionBeginEvent = true
+	case replication.QUERY_EVENT:
+		if !r.everMetGTIDEvent {
+			// user may change session level binlog-format=statement, but it's an unusual operation, so we parse it every time
+			// In MySQL 5.6.x without GTID, the timestamp of BEGIN is not the timestamp of the transaction,
+			// To simplify implementation, we use timestamp of BEGIN as the transaction timestamp,
+			// but this may cause some transaction with timestamp >= target timestamp be skipped.
```

Comment:
> but this may cause some transaction with timestamp >= target timestamp be skipped.

is this caused by GTID_EVENT and QUERY_EVENT has different timestamp? for example GTID_EVENT has ts = 10, QUERY_EVENT has ts = 12. When we input ts = 11, QUERY_EVENT is skipped because of `if !r.everMetGTIDEvent`

Reply:
this case may see In MySQL 5.6.x without GTID, timestamp of BEGIN = timestamp of the first statement, such as:
Query BEGIN ts=1
TableMap ts=1
Rows ts=1
Rows ts=5
Xid ts=10

suppose target ts=10, this transaction is skipped in current impl.